modules/apigee/main.tf

```diff
@@ -117,6 +118,9 @@ resource "google_apigee_instance_attachment" "instance_attachments" {
   instance_id = google_apigee_instance.instances[each.value.region].id
   environment = try(google_apigee_environment.environments[each.value.environment].name,
   "${local.org_id}/environments/${each.value.environment}")
+  lifecycle {
```

Comment:
why is `ignore_changes` needed here?

Reply:
There is an issue which is occurring because `try` function is being used for setting the name of the Apigee Environment. Because of this, the actual value for the `environment` argument cannot be determined until runtime and `terraform plan` shows `force replaces` comment for corresponding Apigee Environment. `terraform apply` force replaces the resource as well. If the same is not fixed, it will end up causing outage on each `terraform apply`. 

`lifecycle ignore_changes` configuration has been added to overcome the issue.

Reply:
I had some doubts about this too:
- why do we need `try` here (to make `destroy` work? because otherwise it should always evaluate properly)
- is indeed `try` the issue here, and not the `for_each` that does the transformations

But in the end it will work as expected, because if you change the environment name, you change the key in the resource, thus forcing replacement. 

I'd love to have a comment here in the code about one and the other (so save my future-self from thinking this over again)

Reply:
The `try` function is actually not even required here as if `var.environments` is `{}` then the instance attachment resource won't even be getting created.

And on another note, `environment` attribute requires the name of the Apigee Environment and not the id.
So usage of `"${local.org_id}/environments/${each.value.environment}"` is not right and instead it should just be `each.value.environment`. However, since we don't require the `try` function here, `environment = google_apigee_environment.environments[each.value.environment].name` should be sufficient.